File: examples/logging_example.js
// RediORM Logging Example
// Run with: redi-orm run logging_example.js

const { fromUri, createLogger, LogLevel } = require('redi/orm');

async function main() {
    // Create a logger (default level is INFO)
    const logger = createLogger('RediORM');
    
    // Set to DEBUG to see SQL queries (optional)
    logger.setLevel(LogLevel.DEBUG);
    
    // Create database connection
    const db = fromUri('sqlite://:memory:');
    
    // Set logger on the database instance
    db.setLogger(logger);
    
    await db.connect();
    
    // Define schema
    await db.loadSchema(`
        model User {
            id    Int     @id @default(autoincrement())
            name  String
            email String  @unique
            age   Int
            posts Post[]
        }
        
        model Post {
            id        Int     @id @default(autoincrement())
            title     String
            content   String
            published Boolean @default(false)
            authorId  Int
            author    User    @relation(fields: [authorId], references: [id])
        }
    `);
    
    // Sync schemas - logs CREATE TABLE statements
    console.log('=== Syncing schemas ===');
    await db.syncSchemas();
    
    // Insert users - logs INSERT statements
    console.log('\n=== Creating users ===');
    const user1 = await db.models.User.create({
        data: {
            name: 'John Doe',
            email: 'john@example.com',
            age: 30
        }
    });
    
    const user2 = await db.models.User.create({
        data: {
            name: 'Jane Smith',
            email: 'jane@example.com',
            age: 25
        }
    });
    
    // Create posts - logs INSERT with foreign keys
    console.log('\n=== Creating posts ===');
    await db.models.Post.create({
        data: {
            title: 'First Post',
            content: 'Hello World',
            authorId: user1.id
        }
    });
    
    // Query with joins - logs SELECT with JOIN
    console.log('\n=== Querying with relations ===');
    const users = await db.models.User.findMany({
        where: { age: { gt: 20 } },
        include: { posts: true },
        orderBy: { name: 'asc' }
    });
    
    console.log(`Found ${users.length} users`);
    
    // Update - logs UPDATE statement
    console.log('\n=== Updating user ===');
    await db.models.User.update({
        where: { id: user1.id },
        data: { age: 31 }
    });
    
    // Aggregation - logs GROUP BY query
    console.log('\n=== Aggregation query ===');
    const result = await db.models.User.aggregate({
        _count: { id: true },
        _avg: { age: true },
        _max: { age: true },
        _min: { age: true }
    });
    
    console.log('Aggregation result:', result);
    
    // Raw query - logs raw SQL
    console.log('\n=== Raw SQL query ===');
    const rawResult = await db.queryRaw('SELECT COUNT(*) as count FROM users WHERE age > ?', 24);
    console.log('Users older than 24:', rawResult[0].count);
    
    // Transaction - logs BEGIN, COMMIT
    console.log('\n=== Transaction ===');
    await db.transaction(async (tx) => {
        await tx.models.Post.create({
            data: {
                title: 'Transaction Post',
                content: 'Created in transaction',
                authorId: user2.id
            }
        });
        
        await tx.models.User.update({
            where: { id: user2.id },
            data: { age: 26 }
        });
    });
    
    console.log('\n=== Example completed! ===');
    await db.close();
}

// Run the example
main().catch(console.error);

File: examples/logging_example.go
package main

import (
	"context"
	"log"
	"os"

	"github.com/rediwo/redi-orm/database"
	_ "github.com/rediwo/redi-orm/drivers/sqlite"
	"github.com/rediwo/redi-orm/schema"
	"github.com/rediwo/redi-orm/types"
	"github.com/rediwo/redi-orm/utils"
)

func main() {
	// Create a logger that outputs to file
	file, err := os.Create("database.log")
	if err != nil {
		log.Fatal("Failed to create log file:", err)
	}
	defer file.Close()

	fileLogger := utils.NewDefaultLogger("RediORM")
	fileLogger.SetLevel(utils.LogLevelDebug)
	fileLogger.SetOutput(file)

	// Create database with logging enabled
	db, err := database.NewFromURI("sqlite://:memory:")
	if err != nil {
		log.Fatal(err)
	}

	// Set the file logger on the database
	db.SetLogger(fileLogger)

	ctx := context.Background()

	// Connect to database
	if err := db.Connect(ctx); err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	// Register a schema
	userSchema := schema.New("User").
		AddField(schema.Field{Name: "id", Type: schema.FieldTypeInt, PrimaryKey: true, AutoIncrement: true}).
		AddField(schema.Field{Name: "name", Type: schema.FieldTypeString}).
		AddField(schema.Field{Name: "email", Type: schema.FieldTypeString, Unique: true}).
		AddField(schema.Field{Name: "age", Type: schema.FieldTypeInt})

	if err := db.RegisterSchema("User", userSchema); err != nil {
		log.Fatal(err)
	}

	// Sync schemas - this will log CREATE TABLE SQL
	if err := db.SyncSchemas(ctx); err != nil {
		log.Fatal(err)
	}

	// Insert a user - this will log INSERT SQL
	user := map[string]any{
		"name":  "John Doe",
		"email": "john@example.com",
		"age":   30,
	}

	result, err := db.Model("User").Insert(user).Exec(ctx)
	if err != nil {
		log.Fatal(err)
	}
	log.Printf("Inserted user with ID: %d\n", result.LastInsertID)

	// Query users - this will log SELECT SQL
	var users []map[string]any
	selectQuery := db.Model("User").Select("id", "name", "email")
	selectQuery = selectQuery.WhereCondition(selectQuery.Where("age").GreaterThan(25))
	selectQuery = selectQuery.OrderBy("name", types.ASC)
	err = selectQuery.FindMany(ctx, &users)

	if err != nil {
		log.Fatal(err)
	}

	log.Printf("Found %d users\n", len(users))

	// Update user - this will log UPDATE SQL
	updateQuery := db.Model("User").Update(map[string]any{"age": 31})
	updateQuery = updateQuery.WhereCondition(updateQuery.Where("email").Equals("john@example.com"))
	updateResult, err := updateQuery.Exec(ctx)

	if err != nil {
		log.Fatal(err)
	}
	log.Printf("Updated %d rows\n", updateResult.RowsAffected)

	// Raw query - this will also be logged
	var count int64
	err = db.Raw("SELECT COUNT(*) FROM users WHERE age > ?", 25).FindOne(ctx, &count)
	if err != nil {
		log.Fatal(err)
	}
	log.Printf("Users older than 25: %d\n", count)

	// Transaction with logging
	err = db.Transaction(ctx, func(tx types.Transaction) error {
		// All operations inside transaction will be logged
		_, err := tx.Model("User").Insert(map[string]any{
			"name":  "Jane Doe",
			"email": "jane@example.com",
			"age":   28,
		}).Exec(ctx)
		return err
	})

	if err != nil {
		log.Fatal(err)
	}

	log.Println("Example completed successfully!")
}
